src/framework/mocha/Utilities/MOJSBlock.m

```diff
@@ -113,13 +113,26 @@ static JSValueRef jsInvoke(MOJavaScriptObject* function, NSMethodSignature* sign
         }
     }
     
-    JSContextRef ctx = [function JSContext];
-    Mocha *runtime = [Mocha runtimeWithContext:ctx];
+    __block JSValueRef value = NULL;
     
-    JSValueRef value = [runtime callJSFunction:[function JSObject] withArgumentsInArray:arguments];
+    void (^execBlock)(void) = ^{
+        JSContextRef ctx = [function JSContext];
+        Mocha *runtime = [Mocha runtimeWithContext:ctx];
+        
+        value = [runtime callJSFunction:[function JSObject] withArgumentsInArray:arguments];
+        
+        if (value == NULL) {
+            [NSException raise:@"CallbackError" format:@"Error while calling the callback"];
+        }
+    };
     
-    if (value == NULL) {
-        [NSException raise:@"CallbackError" format:@"Error while calling the callback"];
+    // Mocha expects all JS functions to be executed on the main thread.
+    // So when creating a block, the JS function still needs to be executed on
+    // the main thread, regardless of where the block is executed.
+    if ([NSThread isMainThread]) {
+        execBlock();
+    } else {
+        dispatch_sync(dispatch_get_main_queue(), execBlock);
```

Comment:
This is generally a code smell that we're writing code without full understanding the circumstances it runs under, but if you're sure it's the best that can be done, so be it.

Reply:
That's a generic block that can then be used anywhere an API expect a block so no real way to know where it will be run no